### PR TITLE
hermes: add more versions, variants and depend on hermes-shm

### PR DESCRIPTION
**Pull request by hyoklee**
See also #46601.

### CODE REVIEW

**Comment by hyoklee** [comment]
@spackbot run pipeline

**Comment by hyoklee** [comment]
@spackbot run pipeline

File: var/spack/repos/builtin/packages/hermes/package.py
Comment (bernhardkaindl) [inline var/spack/repos/builtin/packages/hermes/package.py]: @lukemartinlogan, you are the maintainer of `hermes`, any review from you?

Trying to build this, I saw that does not work (with hermes@1.2.1):

When -DHERMES_ENABLE_MPIIO_ADAPTER=ON is defined, and the none of the other two is defined, the build files with:
```py
spack-src/hermes_adapters/mpiio/mpiio_fs_api.h:696:2: error: #error "No MPI implementation specified for MPIIO adapter"
```
This is the spec `hermes@1.2.1%gcc@13.1.0~adios~ares~compress~encrypt~ipo+mpiio~python+stdio~vfd~zmq`

It only has `+mpiio`, but does not have `openmpi` or `mpich` because those are not variants of `hermes` and you cannot add them to the spec without that, AFAIK.

AFAICS, You need to select which MPI to compile with, and of course `hermes-shm+mpiio` should use the same MPI. Apparently, this seems to be a way to get the MPI that is selected using the "mpi" alias:
```py
var/spack/repos/builtin/packages/r-rmpi/package.py:        mpi_name = spec["mpi"].name
var/spack/repos/builtin/packages/r-rmpi/package.py:        if mpi_name == "openmpi":
var/spack/repos/builtin/packages/r-rmpi/package.py:        elif mpi_name == "mpich":
g
```

File: var/spack/repos/builtin/packages/hermes/package.py
Comment (bernhardkaindl) [inline var/spack/repos/builtin/packages/hermes/package.py:54]:
```suggestion
    depends_on("hermes-shm+mpiio", when="+mpiio")
    depends_on("mpi", when="+mpiio")
```
This is needed to get the `mpi` implementation into the build.
Without it, no `mpi` is added. Needed to get `spec["mpi"].name` as well

You also use the same line in `hermes-shm`, so you should use the same mechanism here too.

**Review by bernhardkaindl — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/hermes/package.py
Comment (bernhardkaindl) [inline var/spack/repos/builtin/packages/hermes/package.py:37]: As commented and as also used in `hermes-shm`, this shall not be removed, but needs to be replaced with `depends_on("mpi", when="+mpiio")`.

**Review by bernhardkaindl — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/hermes/package.py
Comment (bernhardkaindl) [reply var/spack/repos/builtin/packages/hermes/package.py]: I found that `spec["..."].name` is also used for getting the selected `blas` in other recipes.

Pushed an initial attempt.

But as it needs a further fix, I'll mark this as draft until ready to prevent GitLab CI from spending paid compute Pods on building it.

File: var/spack/repos/builtin/packages/hermes/package.py
Comment (bernhardkaindl) [inline var/spack/repos/builtin/packages/hermes/package.py:61]: `depends_on("mpi")` -> `depends_on("mpi", when="+mpiio")` - Is needed to get the selected virtual for `"mpi"` in spack.

The conflict with `nvhpc` is needed as the selection which, variant to use does not have support for `"mpi" from a build of `nvhpc+mpi`, and if one such build already exists, and none for `openmpi` or `mpich`, the `concretizer` does select `nvhpc` for `"mpi"` if  `--reuse` (default) is in effect.

File: var/spack/repos/builtin/packages/hermes/package.py
Comment (bernhardkaindl) [inline var/spack/repos/builtin/packages/hermes/package.py:73]: This now takes care of getting the selected virtual for `"mpi"` (as used in other recipes for `mpi` and `blas`) and raises an install error if another provider that is not supported is selected.

File: var/spack/repos/builtin/packages/hermes-shm/package.py
Comment (bernhardkaindl) [inline var/spack/repos/builtin/packages/hermes-shm/package.py:43]: Tested: The current versions of `doxygen` are compatible with this older versions.
Build also succeeds with `doxygen@1.11.0`, hence it does not need to be pinned to a single version.

**Review by bernhardkaindl — DISMISSED** [review]

File: var/spack/repos/builtin/packages/hermes/package.py
Comment (bernhardkaindl) [inline var/spack/repos/builtin/packages/hermes/package.py]: Just as a question (just curious): Do you need all those versions?

If there is no requirement to have the checksums of all these versions, I would not add them, because versions accumulate anyway, and once you have versions listed, you'd have to deprecate them before removing them again to limit the size of the recipe.
```suggestion
```

**Review by bernhardkaindl — APPROVED** [review]
Build results on Ubuntu 22.04.4 LTS:
```py
bin/spack find --variants adios2 hermes-shm hermes~mpiio~stdio+adios@master hermes~mpiio~stdio+adios@1.2.1 hermes~mpiio~stdio+ares@master hermes~mpiio~stdio+ares@1.2.1 hermes~mpiio~stdio+compress@master hermes~mpiio~stdio+compress@1.2.1 hermes~mpiio~stdio+encrypt@master hermes~mpiio~stdio+encrypt@1.2.1 hermes~stdio@master hermes~stdio@1.2.1 hermes~mpiio~stdio+python@master hermes~mpiio@master hermes~mpiio@1.2.1 hermes~mpiio~stdio+zmq@master hermes~mpiio~stdio+zmq@1.2.1 libcatalyst
-- linux-ubuntu22.04-skylake / gcc@13.1.0 -----------------------
adios2@2.10.1+blosc2+bzip2+fortran+libcatalyst+mgard+mpi+png+sst+sz+zfp patches=4247a85
hermes@1.2.1
hermes@1.2.1+zmq
hermes@1.2.1+encrypt
hermes@1.2.1+compress
hermes@1.2.1+ares
hermes@1.2.1+adios
hermes@master
hermes@master+zmq
hermes@master+python
hermes@master+encrypt
hermes@master+compress
hermes@master+ares
hermes@master+adios
hermes-shm@master+boost+cereal+mochi+mpiio
hermes-shm@master+boost+cereal+mochi+mpiio+zmq
hermes-shm@master+boost+cereal+encrypt+mochi+mpiio
hermes-shm@master+boost+cereal+compress+mochi+mpiio
hermes-shm@master+ares+boost+cereal+mochi+mpiio
hermes-shm@master+adios+boost+cereal+mochi+mpiio
libcatalyst@2.0.0
```

## One failed spec:
- `hermes~mpiio~stdio+python@1.2.1`
### `hermes~mpiio~stdio+python@1.2.1`:
```py
==> Error: failed to concretize `hermes@1.2.1~mpiio+python~stdio` for the following reasons:
     1. Cannot select a single "version" for package "hermes"
     2. Cannot satisfy 'hermes@master'
     3. Cannot satisfy 'hermes@1.2.1'
     4. Cannot satisfy 'hermes@1.2.1'
        required because hermes@1.2.1~mpiio+python~stdio requested explicitly
```
spack failed to concretize the spec due to a conflict.
This is likely intentional due to a conflict() in the recipe(s).

Generated by https://github.com/spack/build-quality-tools/blob/main/build_pr_changes.py
```py
/home/bkaindl/build-quality-tools/build_pr_changes.py --approve
```

**Comment by hyoklee** [comment]
Thank you @bernhardkaindl so much for improving this PR!

**Comment by bernhardkaindl** [comment]
@hyoklee If you agree with these changes, ask spackbot to run the GitLab pipeline:

> @spackbot run pipeline

This should then schedule running the GitLab pipeline, and it can be merged after that.
You can mention me using @bernhardkaindl when the GitLab pipeline was successful and all checks have passed for merging it.

**Comment by hyoklee** [comment]
@spackbot run pipeline

**Comment by bernhardkaindl** [comment]
One job failure, error not visible: https://gitlab.spack.io/spack/spack/-/jobs/12952078
Try it again:
@spackbot run pipeline

**Comment by bernhardkaindl** [comment]
Hm, that did nothing, retrying with re-run:
@spackbot re-run pipeline

**Review by bernhardkaindl — APPROVED** [review]
Rerunning the pipelines did not solve the CI failure, it repeated again.

I rebased the PR on top of the latest GitLab CI commit on `develop`.
I hope this fixes the CI now.

**Comment by hyoklee** [comment]
@bernhardkaindl , thank you so much running CI tests again! 

All checks passed!

**Pull request merged**
